### ISSUE TITLE
[stable/memcached]: defaults to 3 replicas, but no data is shared between them

### ISSUE THREAD

**Issue by plumdog** [helm/charts#11773]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Bug report

**Version of Helm and Kubernetes**:
Helm: 2.10
Kubernetes: 1.10

**Which chart**:
stable/memcached

**What happened**:
Installed with defaults, gave me a stateful set with 3 replicas. Setting a value works, but only sets the value in one of the replicas. There appears to be no replication of data between the replicas. This means that they're actually 3 separate memcached servers, and it depends on DNS as to which one you get, so this is, I think, worse than just having 1.

**What you expected to happen**:
Either:
- data to be replicated between the pods somehow
    - either by letting them discover each other/leader election/other (I have no idea if this is something memcache knows how to do this.)
    - or by sharing a volume, maybe, and some docs to tell you that if you want it to be HA, you need a volume that supports this. (I have no idea if memcache can share data this way. I suspect not.)
- the chart to default to one pod, and explain that having multiple pods is probably not what you want.

**How to reproduce it** (as minimally and precisely as possible):

```bash
$ helm install stable/memcached --name memcached-test --namespace test
$ kubectl -n test run debug --image=ubuntu --command sleep 10000
$ kubectl -n test exec -it debug-abc-123 bash
---
$ apt install dnsutils netcat
# set using the service
$ echo -e 'set mykey 0 60 5\r\nhello\r\nquit\r\n' | nc memcached-test.test.svc.cluster.local 11211
STORED 
# get using the pod IPs
$ for podip in $(dig +short memcached-test.test.svc.cluster.local); do echo pod ip: $podip; echo -e 'get mykey\r\nquit\r\n' | nc $podip 11211; done
pod ip: 100.96.1.174
END
pod ip: 100.96.3.5
VALUE mykey 0 5
hello
END
pod ip: 100.96.2.2
END
# See that it has only been stored in one of the pods
```

**Anything else we need to know**:

I think the correct course of action is probably:
1) make it so the memcached chart defaults to 1 replica, to remove this surprising behaviour, and put a note about it in the documentation (see #11774).
2) someone who understands memcached to investigate whether it is possible to make these pods actually be replicas. (I will make a start on this, but I am starting from scratch)

Also, this behaviour would appear to have existed by default for a while. It was introduced here: https://github.com/helm/charts/pull/933/files#diff-e439a54f94b1162c00cdd40f9b6f8123R14

Pinging @olemarkus @gtaylor @kennethaasan because they are the owners of the memcached chart, and @dhilipkumars because they made #933.

**Comment by timothyclarke** [comment]
Isn't lack of replication a standard memcached thing ?
If you want replication you either do it in the client, or use recached

**Comment by plumdog** [comment]
@timothyclarke That's sort of my point, really. If memcached doesn't do replication, why does it default to 3 "replicas"?

**Comment by plumdog** [comment]
Pinging @olemarkus and @kennethaasan because they're assigned to the associated PR.

**Comment by olemarkus** [comment]
There are some things to consider here:
* how well does an app respond to zero cache backends being available?
* how well does an app respond to _all_ cache being gone?

You'd typically use consistent hashing across these servers so that 1 cache entry is on one single pod.

**Comment by plumdog** [comment]
@olemarkus I agree that something dropping to zero cache backends being available will result in errors, but at least they will be easily understood errors, rather than strange "set value=foo, update value=bar, get value and find foo" behaviour.

@dhilipkumars But am I right that there's nothing to make these replicas behave like shards though?

You both refer to the idea of sharding the keys over the different pods to distribute load, but there's nothing in the chart to handle anything like that. From a quick google, adding that would require some proxy to route requests to the right pods and is not something that memcached natively understands. If sharding were simple to add in front of the replicas, it would mean consistent behaviour, but it would not mean HA (so would be very roughly similar to 1 pod 3 times the size). There's also other questions with sharding, eg: what happens if you change the size of the statefulset? I don't think sharding is the solution to the problem here.

I think the only argument in favour of keeping the replica count > 1 is that it means that we have an iffy kind of HA, insofar as there would be a memcached for it to connect to. But I think it will behave so strangely that it is not the right thing to do by default (imagine trying to delete something from your cache, but it still being there, but only sometimes depending on which backend the k8s service connected you to). If someone needed a memcached that was up, and didn't mind about this strange behaviour, they could ask for multiple replicas (this PR does just set the default behaviour, after all) and in my PR I have added a bit to the readme about doing this. But I think the default behaviour should be straightforward and predictable.

TLDR: the default values should be simple and unsurprising.

**Comment by olemarkus** [comment]
A memcache server doesn't know anything about other servers. So sharing is a feature implemented by the clients. The feature you want to google is called consistent hashing. One single key would then only exist on one pod.

**Comment by plumdog** [comment]
If the client needs to know about the different pods, them hiding behind a service means the client can't differentiate. You'd need 1 service per pod, I think.

I'd suggest the way to handle the situation you describe is to use this chart multiple times, with replicas=1.

Also, my goal is not to make the chart use sharding, only to make it so the default behaviour is sensible.

**Comment by olemarkus** [comment]
It is behind a headless service, so there is no proxy/loadbalancer. In some cases clients requires that you enter the deterministic (since this is a statefulset) hostnames.

The default is sensible. Having only one replica for a prod setup is not recommended for the reasons stated earlier.
